File: examples/python/basic-langchain-agent/src/middleware.py
"""
LangChain middleware for logging tool calls.
"""

from langchain.agents.middleware import AgentMiddleware

from langchain.tools.tool_node import ToolCallRequest
from typing import Callable, Any


class ToolCallMiddleware(AgentMiddleware):
    """Middleware that logs tool call IDs."""

    def wrap_tool_call(
        self,
        request: ToolCallRequest,
        handler: Callable[[ToolCallRequest], Any],
    ) -> Any:
        print(f"[Middleware] Tool call ID: {request.tool_call['id']}\n")
        result = handler(request)
        return result

File: examples/typescript/basic-langchain-agent/src/middleware.ts
import {
  createMiddleware,
  type ToolCallHandler,
  type ToolCallRequest,
} from 'langchain'

export const toolCallMiddleware = createMiddleware({
  name: 'ToolCallMiddleware',
  wrapToolCall: async (
    request: ToolCallRequest<any, any>,
    handler: ToolCallHandler,
  ) => {
    console.log(`[Middleware] Tool call ID: ${request.toolCall.id}\n`)
    const result = await handler(request)
    return result
  },
})
